### PR TITLE
Supervised process cleanup function

### LINKED ISSUE #2218: Konnectivity server unclean shutdown & failure to re-start
### Before creating an issue, make sure you've checked the following:

- [X] You are running the latest released version of k0s
- [X] Make sure you've searched for existing issues, both open and closed
- [X] Make sure you've searched for PRs too, a fix might've been merged already
- [X] You're looking at docs for the released version, "main" branch docs are usually ahead of released versions.

### Platform

```shell
Linux 5.15.0-48-generic #54-Ubuntu SMP Fri Aug 26 13:26:29 UTC 2022 x86_64 GNU/Linux
PRETTY_NAME="Ubuntu 22.04.1 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.1 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

### Version

v1.25.2+k0s.0

### Sysinfo

<details><summary>k0s sysinfo</summary>
```text
Machine ID: "3fdd4c5daac60aaf51999e3bfa8693f56006e086534e5b87852c7f69af3cd4aa" (from machine) (pass)
Total memory: 969.5 MiB (warning: 1.0 GiB recommended)
Disk space available for /var/lib/k0s: 8.1 GiB (pass)
Operating system: Linux (pass)
  Linux kernel release: 5.15.0-48-generic (pass)
  Max. file descriptors per process: current: 1048576 / max: 1048576 (pass)
  Executable in path: modprobe: /usr/sbin/modprobe (pass)
  /proc file system: mounted (0x9fa0) (pass)
  Control Groups: version 2 (pass)
    cgroup controller "cpu": available (pass)
    cgroup controller "cpuacct": available (via cpu in version 2) (pass)
    cgroup controller "cpuset": available (pass)
    cgroup controller "memory": available (pass)
    cgroup controller "devices": available (assumed) (pass)
    cgroup controller "freezer": available (assumed) (pass)
    cgroup controller "pids": available (pass)
    cgroup controller "hugetlb": available (pass)
    cgroup controller "blkio": available (via io in version 2) (pass)
  CONFIG_CGROUPS: Control Group support: built-in (pass)
    CONFIG_CGROUP_FREEZER: Freezer cgroup subsystem: built-in (pass)
    CONFIG_CGROUP_PIDS: PIDs cgroup subsystem: built-in (pass)
    CONFIG_CGROUP_DEVICE: Device controller for cgroups: built-in (pass)
    CONFIG_CPUSETS: Cpuset support: built-in (pass)
    CONFIG_CGROUP_CPUACCT: Simple CPU accounting cgroup subsystem: built-in (pass)
    CONFIG_MEMCG: Memory Resource Controller for Control Groups: built-in (pass)
    CONFIG_CGROUP_HUGETLB: HugeTLB Resource Controller for Control Groups: built-in (pass)
    CONFIG_CGROUP_SCHED: Group CPU scheduler: built-in (pass)
      CONFIG_FAIR_GROUP_SCHED: Group scheduling for SCHED_OTHER: built-in (pass)
        CONFIG_CFS_BANDWIDTH: CPU bandwidth provisioning for FAIR_GROUP_SCHED: built-in (pass)
    CONFIG_BLK_CGROUP: Block IO controller: built-in (pass)
  CONFIG_NAMESPACES: Namespaces support: built-in (pass)
    CONFIG_UTS_NS: UTS namespace: built-in (pass)
    CONFIG_IPC_NS: IPC namespace: built-in (pass)
    CONFIG_PID_NS: PID namespace: built-in (pass)
    CONFIG_NET_NS: Network namespace: built-in (pass)
  CONFIG_NET: Networking support: built-in (pass)
    CONFIG_INET: TCP/IP networking: built-in (pass)
      CONFIG_IPV6: The IPv6 protocol: built-in (pass)
    CONFIG_NETFILTER: Network packet filtering framework (Netfilter): built-in (pass)
      CONFIG_NETFILTER_ADVANCED: Advanced netfilter configuration: built-in (pass)
      CONFIG_NETFILTER_XTABLES: Netfilter Xtables support: module (pass)
        CONFIG_NETFILTER_XT_TARGET_REDIRECT: REDIRECT target support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_COMMENT: "comment" match support: module (pass)
        CONFIG_NETFILTER_XT_MARK: nfmark target and match support: module (pass)
        CONFIG_NETFILTER_XT_SET: set target and match support: module (pass)
        CONFIG_NETFILTER_XT_TARGET_MASQUERADE: MASQUERADE target support: module (pass)
        CONFIG_NETFILTER_XT_NAT: "SNAT and DNAT" targets support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: "addrtype" address type match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_CONNTRACK: "conntrack" connection tracking match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_MULTIPORT: "multiport" Multiple port match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_RECENT: "recent" match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_STATISTIC: "statistic" match support: module (pass)
      CONFIG_NETFILTER_NETLINK: module (pass)
      CONFIG_NF_CONNTRACK: Netfilter connection tracking support: module (pass)
      CONFIG_NF_NAT: module (pass)
      CONFIG_IP_SET: IP set support: module (pass)
        CONFIG_IP_SET_HASH_IP: hash:ip set support: module (pass)
        CONFIG_IP_SET_HASH_NET: hash:net set support: module (pass)
      CONFIG_IP_VS: IP virtual server support: module (pass)
        CONFIG_IP_VS_NFCT: Netfilter connection tracking: built-in (pass)
      CONFIG_NF_CONNTRACK_IPV4: IPv4 connetion tracking support (required for NAT): unknown (warning)
      CONFIG_NF_REJECT_IPV4: IPv4 packet rejection: module (pass)
      CONFIG_NF_NAT_IPV4: IPv4 NAT: unknown (warning)
      CONFIG_IP_NF_IPTABLES: IP tables support: module (pass)
        CONFIG_IP_NF_FILTER: Packet filtering: module (pass)
          CONFIG_IP_NF_TARGET_REJECT: REJECT target support: module (pass)
        CONFIG_IP_NF_NAT: iptables NAT support: module (pass)
        CONFIG_IP_NF_MANGLE: Packet mangling: module (pass)
      CONFIG_NF_DEFRAG_IPV4: module (pass)
      CONFIG_NF_CONNTRACK_IPV6: IPv6 connetion tracking support (required for NAT): unknown (warning)
      CONFIG_NF_NAT_IPV6: IPv6 NAT: unknown (warning)
      CONFIG_IP6_NF_IPTABLES: IP6 tables support: module (pass)
        CONFIG_IP6_NF_FILTER: Packet filtering: module (pass)
        CONFIG_IP6_NF_MANGLE: Packet mangling: module (pass)
        CONFIG_IP6_NF_NAT: ip6tables NAT support: module (pass)
      CONFIG_NF_DEFRAG_IPV6: module (pass)
    CONFIG_BRIDGE: 802.1d Ethernet Bridging: module (pass)
      CONFIG_LLC: module (pass)
      CONFIG_STP: module (pass)
  CONFIG_EXT4_FS: The Extended 4 (ext4) filesystem: built-in (pass)
  CONFIG_PROC_FS: /proc file system support: built-in (pass)
```

</details>

### What happened?

After running k0s controller for some time, we experience that the konnectivity server backend is offline and won't recover. When we inspect the logs, we see that k0s fails to start konnectivity-server. 

We are experience this since switch to `v1.25.2+k0s.0`. We are running 3 controllers and external haproxy. 
The only way to reliably recover from that situation is fully reboot the server. Sometimes removing the unix socket also works. After that, konnectivity server is online for some time, but then goes offline after some time and the problem re-appears

### Steps to reproduce

1. Let controller running for a while

### Expected behavior

k0s should be able to recover from that situation and start konnectivity server (either properly cleanup after crash, or re-bind to existing socket if possible) 

### Actual behavior

k0s failes to start konnectivity server

### Screenshots and logs

```
Oct 03 13:44:41 xxxx k0s[22944]: time="2022-10-03 13:44:41" level=info msg="Error: failed to run the frontend server: failed to get uds listener: failed to listen(unix) name /run/k0s/konnectivity-server/konnectivity-server.sock: listen unix /run/k0s/konnectivity-server/konnectivity-server.sock: bind: address already in use" component=konnectivity
[...]
Oct 03 13:44:41 xxxx k0s[22944]: time="2022-10-03 13:44:41" level=warning msg="exit status 1" component=konnectivity
Oct 03 13:44:41 xxxx k0s[22944]: time="2022-10-03 13:44:41" level=info msg="respawning in 5s" component=konnectivity

[repeat endlessly]
```

```
$ ls -lsa /run/k0s/konnectivity-server/konnectivity-server.sock
0 srwxrwx--- 1 konnectivity-server root 0 Oct  3 09:10 /run/k0s/konnectivity-server/konnectivity-server.sock
```

```
$ lsof /run/k0s/konnectivity-server/konnectivity-server.sock

```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by makhov**
Signed-off-by: Alexey Makhov <amakhov@mirantis.com>

## Description

Fixes #2218

A function to clean leftovers before starting or restarting the supervised process is added to Supervisor API

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [ ] Manual test
- [ ] Auto test added

## Checklist:

- [ ] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [ ] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by jnummelin** [comment]
For the konnectivity-server it also has this flag available:
```
--delete-existing-uds-file           If true and if file UdsName already exists, delete the file before listen on that UDS file
```
Setting that would solve the konnectivity-server issue. But there might be other cases where we need cleanup too so it does make sense to have a generic solution to the problem too.

**Comment by makhov** [comment]
@jnummelin thanks! I've added the flag to the arguments, but keep the `CleanBeforeFn` in the API.

File: pkg/supervisor/supervisor.go [hunk 1]
```diff
@@ -135,6 +137,13 @@ func (s *Supervisor) Supervise() error {
 		restarts := 0
 		for {
 			s.mutex.Lock()
+
+			if s.CleanBeforeFn != nil {
+				err := s.CleanBeforeFn()
+				if err != nil {
+					s.log.Warnf("Failed to clean before running the process %s: %s", s.BinPath, err)
```
Comment (soider) [inline pkg/supervisor/supervisor.go:144]: I think if there is a cleanup function and if it fails it might have sense to fail supervisor as well

Comment (makhov) [reply pkg/supervisor/supervisor.go:144]: It's running in a goroutine, so it'll just stop supervising

**Review by soider — APPROVED** [review]

**Pull request merged**